Issue #14697: Resolved False negative in UnusedImportCheck

zyadahmed opened the pull request:
#14697 

The issue was  `@link` tag being considered valid. The root cause lies in the **`BLOCK_TAG_PATTERN`**, which retrieves all regex expressions containing `@` within Javadoc comments. This pattern incorrectly identifies `@link` as an valid reference for imports.

https://github.com/checkstyle/checkstyle/blob/1fa0377d6892f12e544e7c56acec4cf38c23d15e/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/utils/BlockTagUtil.java#L41
To resolve this, we filter patterns with `@link` since they are considered invalid references for imports.

---

Another issue arises with `{@link Set::add}`, which is an invalid link. correct one `{@link ClassName#methodName}`

The probelm was in extraction of  INLINE_TAG_PATTERN regex  
https://github.com/checkstyle/checkstyle/blob/1fa0377d6892f12e544e7c56acec4cf38c23d15e/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/utils/InlineTagUtil.java#L39

zyadahmed commented:
@romani is `@see` and `@linkplain `consider import usage in checkstyle or not ?

zyadahmed commented:
> @romani is `@see` and `@linkplain `consider import usage in checkstyle or not ?

@romani

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -344,11 +346,21 @@ private void collectReferencesFromJavadoc(DetailAST ast) {
+        tags.addAll(
+                getValidTags(textBlock, JavadocUtil.JavadocTagType.BLOCK)
+                        .stream()
+                        .filter(tag -> !tag.getTagName().equals(JavadocTagInfo.LINK.getName()))
```
this exclude by name only is suspicious.
we should do somethig like isValidMethodLinkTag, some general method that can validate tag link for validnes.

please explain this  a bit more.

romani replied:
do we mean to say that getValidTags should not return LINK tag ?

romani replied:
sounds like getValidTags should do better job to do this filtering for both cases.

zyadahmed replied:
> do we mean to say that getValidTags should not return LINK tag ?

yes becasue it return all tags similar like that @link @see and this format is not valid so we exclude it the only valid format for link should be between {}

romani replied:
Let's move logic of filtering to getValidTags

romani replied:
Let's keep all mess in Check.
Method should be until only if it proven to be good for all usages

romani reviewed (changes requested):
in general all good,

items to clarify:

romani commented:
https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html
Search for online tags:
{@link package.class#member label}
{@linkplain package.class#member label}

So filtering should be named as `isBlockTag` that filters out all inline tags.

Read on difference of block and inline tags at `Block and inline Tags` chapter of https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html

romani commented:
@mahfouz72 , please look at this PR, to get understanding of how diverse javadoc .

zyadahmed commented:
@romani I changed and moved all logic to JavadocUtil to filter all tags from the starting point and resolved all mutations related to new logic . However, the CI failed because other mutations failed in different parts of the code.

mahfouz72 commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -130,6 +130,21 @@ public static JavadocTags getJavadocTags(TextBlock textBlock,
+     * @param javadocTag the type of the Javadoc tag
+     * @return {@code true} if the tag is valid, otherwise {@code false}
+     */
+    public static boolean isValidTag(TagInfo tag, JavadocTagType javadocTag) {
+        boolean isValid = true;
+        if (javadocTag == JavadocTagType.BLOCK) {
+            isValid = !tag.getName().equals(JavadocTagInfo.LINK.getName());
+        }
+        return isValid && !tag.getValue().contains("::");
+    }
```
All of these can reference imports

https://github.com/checkstyle/checkstyle/blob/8531fec143058ddd92fab14839f2e9389ac6cf5c/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTag.java#L157-L164
   ---
   
Your method deals with links only. I think direction is good but we have to account for all others tag and do more work.
   
Tags That Reference Imports:

Block Tags: `@see`, `@throws`, `@exception`
Inline Tags: `{@link}`, `{@value}`, `{@linkplain}`

Filtering Logic:

- If the tag type is block (i.e., it does not have `{ }`), then exclude it if its name is link, value, or linkplain (since these are only valid as inline tags).

- Validate the value to ensure it is a correct reference (e.g., no `::` method references, correct format for links).

zyadahmed replied:
> * Validate the value to ensure it is a correct reference (e.g., no `::` method references, correct format for links).

@romani @mahfouz72   The problem is to validate whether a given value is a correct reference. Using the same approach, we need to match it with a regex. However, the regex would be too complex and might not handle some edge cases. What is the correct approach to solve this problem?

romani replied:
it is ok, all validation by regexp has problems on edge cases, name method `bestTryToMatchReference`.

We will never handle all, but we can handle some basic.
When Check is rewritten in AST based model, it will be more strict on validation

mahfouz72 reviewed (changes requested):
item

romani commented:
CI is very red

zyadahmed commented:
> CI is very red

I will try to fix it for some reason related to mutations and regex, but is the solution good?

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -118,18 +130,121 @@ public static JavadocTags getJavadocTags(TextBlock textBlock,
-            if (JavadocTagInfo.isValidName(tag.getName())) {
-                validTags.add(
-                    new JavadocTag(line, col, tag.getName(), tag.getValue()));
+            if (JavadocTagInfo.isValidName(tag.getName()) && isValidImportTag(tag, tagType)) {
```
For `getJavadocTags(TextBlock` it is not good to filter out valid .

Let's move isValidImportTag back to Check.
Please rename to isValidTagWithJavaElemetRegerence

Please avoid "import" in all other names. It is reference it is not import

zyadahmed replied:
@romani But is that considered invalid tags from the beginning without reaching any checks? Because in `JavadocUtil#getJavadocTags`, it actually separates both valid and invalid tags. In that case, do we consider them invalid from the beginning and not treat them as valid tags?

https://github.com/checkstyle/checkstyle/blob/e06aaf2ef8676e6001d9d85e523964477730012a/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java#L88 
as it show here we get from that method valid tags only ?

romani replied:
this is good usage:
https://github.com/checkstyle/checkstyle/blob/cf745474e145584543684f2c812d3eeed31328ff/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTypeCheck.java#L378-L382

decision on valid of not valid is happening in Check, but each Check will see tags to be invalid by their own rules.
If we put more in colletion of invalid to JavadocTags, it might affect other Check.
   
we need a bit more deep refacoting here to put logic on what is valid and what is invalid to Checks. Util class just have functions that are shared between Checks, but Util functions does not make opinion on valid. 
JavadocTypeCheck, UnusedImportsCheck will be affected by this update, it is ok.

as you split/copy logic fromm util  to Checks it will be more logical and each check can decided by itself what is valid.

zyadahmed replied:
Okay, I will work on it. Thanks a lot for the explanation! So, we can fix the problem by adding the filtering here in that part, right? 
https://github.com/checkstyle/checkstyle/blob/cf745474e145584543684f2c812d3eeed31328ff/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java#L355-L358

romani replied:
Yes

romani reviewed (changes requested):
In general ok.

Items

zyadahmed commented:
@romani I tried to make the CI green, but the CI/CircleCI errors are unrelated to the classes I changed. The failure  in pitest test are related to JavadocUtil, from which I removed all my new changes. also checke (index), where I had already suppressed the check error because it was related to how I handle the logic. However, all other checks have passed, and I have resolved all mutations related to my code.

zyadahmed commented:
@romani  i changed all logic in UnusedImportCheck but i have some probelms in ci  in Checker (index) i already suppressed the erros but still not working and i dont have any idea how to solve ci/circleci

rnveach commented:
`no-error-*` runs are against 3rd party repos. It is validating your changes against other sources to see if they will cause regression on other projects. You should be able to look at the CIs to see where the errors come from and look at the actual files. Verify that the new violations being reported are correct and not bugs in your implementation. You may want to look to see if any should be added to our own tests. You should post some examples here for @romani and others to confirm. Once the check updates are resolved here to be all good (someone will most likely tell you when we reach this), you will need to provide fixes to the 3rd party repos for them to merge and CI to pass here.

rnveach commented:
`run-inspections` can be found at https://app.circleci.com/pipelines/github/checkstyle/checkstyle/1/workflows/4ff6fca1-b529-462c-b71e-b4cfdbf47c6f/jobs/821142/artifacts . Let us know if there is some error you are not understanding how to resolve.

`semaphoreci` failure is another no-run error.

romani commented on config/checker-framework-suppressions/checker-index-suppressions.xml:
```diff
@@ -1078,6 +1078,28 @@
+    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java</fileName>
+    <specifier>argument</specifier>
+    <message>incompatible argument for parameter endIndex of String.substring.</message>
+    <lineContent>referenceString = content.substring(0, referenceIndex);</lineContent>
```
while I reviewing you PR,
please share reasons on not fixing this Checker violation.

zyadahmed replied:
> while I reviewing you PR, please share reasons on not fixing this Checker violation.

The extractReferencePart method returns a value ranging from `0` to `input.length()-1 `if the reference is valid,  there is no chance of going out of bounds. In the case of an invalid reference, it returns -1, which is already handled in the parent method. Adding extra handling for this case would introduce unnecessary code, leading to a Pitest suppression without any real user scenario to handle that case

romani replied:
thanks a lot.
Do you know any annotation on variable that would potentially allow to avoid violation from Checker? 
wihtout intruduction of new jar dependecy.

romani replied:
@zyadahmed , ^^

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -344,17 +356,23 @@ private void collectReferencesFromJavadoc(DetailAST ast) {
-            .filter(JavadocTag::canReferenceImports)
-            .forEach(tag -> references.addAll(processJavadocTag(tag)));
+                .filter(JavadocTag::canReferenceImports)
+                .forEach(tag -> references.addAll(processJavadocTag(tag)));
```
please revert unrelated formatting to simplify diff

zyadahmed replied:
> please revert unrelated formatting to simplify diff

done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -367,6 +385,7 @@ private static Set<String> collectReferencesFromJavadoc(TextBlock textBlock) {
      */
     private static List<JavadocTag> getValidTags(TextBlock cmt,
             JavadocUtil.JavadocTagType tagType) {
+
```
revert please

zyadahmed replied:
> revert please

done

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/unusedimports/InputUnusedImportsWithLinkAndMethodTagSuperClass.java:
```diff
@@ -0,0 +1,14 @@
+public class InputUnusedImportsWithLinkAndMethodTagSuperClass {
+        public void Test(LocalDate date, Collection s){}
+        public void setUrl(URI url){}
+        public void extractParameters(double[] d, Month month){
+
+        }
```
`{}`

zyadahmed replied:
> `{}`

done

romani reviewed (changes requested):
shallow review:

romani commented on config/checker-framework-suppressions/checker-regex-property-key-compiler-message-suppressions.xml:
```diff
@@ -297,6 +297,13 @@
+    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java</fileName>
+    <specifier>group.count.unknown</specifier>
+    <message>unable to verify non-literal groups parameter.</message>
+    <lineContent>final String methodPart = matcher.group(methodIndex);</lineContent>
```
please explain this.

romani replied:
https://checkerframework.org/manual/#regex-capturing-groups
https://checkerframework.org/api/org/checkerframework/checker/regex/qual/Regex.html
https://checkerframework.org/manual/#regex-annotations

romani replied:
can something like this help to avoid violation ?
```
$ git diff
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java b/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java
index 9e0ecf6231..0c14541fd7 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java
@@ -29,6 +29,8 @@ import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import org.checkerframework.checker.regex.qual.Regex;
+
 import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
 import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
@@ -134,7 +136,7 @@ public class UnusedImportsCheck extends AbstractCheck {
         CommonUtil.createPattern("^java\\.lang\\.[a-zA-Z]+$");
 
     /** Reference pattern. */
-    private static final Pattern REFERENCE = Pattern.compile(
+    private static final @Regex(3) Pattern REFERENCE = Pattern.compile(
             "^([a-z_$][a-z\\d_$<>.]*)?(#(.*))?$",
             Pattern.CASE_INSENSITIVE
     );

```

```
✔ ~/java/github/romani/checkstyle [zyadahmed/UnusedImportsIssue L|✚ 1⚑ 2] 
$ groovy ./.ci/checker-framework.groovy checker-regex-property-key-compiler-message

```

zyadahmed replied:
I tried this, but it still produces the same error.

romani replied:
Please create issue on them to let them help on this violation and how to resolve it

zyadahmed replied:
> Please create issue on them to let them help on this violation and how to resolve it

can we make it in separate pr after that beacause that pr is too long now or what do you think ?

romani replied:
yes, we can, but please share how it can be fixed.

romani commented on config/checker-framework-suppressions/checker-index-suppressions.xml:
```diff
@@ -1078,6 +1078,28 @@
+  <checkerFrameworkError unstable="false">
+    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java</fileName>
+    <specifier>argument</specifier>
+    <message>incompatible argument for parameter endIndex of String.substring.</message>
+    <lineContent>referenceString = content.substring(0, referenceIndex);</lineContent>
+    <details>
+      found   : int
+      required: @NonNegative int
```
please explain this

zyadahmed replied:
> please explain this

Here, it generates a warning because `extractReferencePart` produces a result within the valid range if it is a valid part  otherwise it returns an invalid index, which in the implementation is -1. However, that case is already handled. I tried to find a better solution, but it would either complicate the implementation or require a different return value that does not align with the method purpose.

romani replied:
yes, agree, code logic enforce it to be positive.
```
        if (referenceIndex != -1) {
            final String referenceString;
            if (referenceIndex == 0) {
                referenceString = content;
            }
            else {
                referenceString = content.substring(0, referenceIndex);
```

zyadahmed replied:
> @zyadahmed , please open issue on checkerframework to get to know if they can improve validation, so this suppression will be temporal until they improve

I opened it and will try to follow along until someone resolves it.

romani replied:
Please share link

zyadahmed replied:
> Please share link

https://github.com/typetools/checker-framework/issues/7008

romani replied:
can we apply what is suggested in that issue.

romani replied:
if `<lineContent>referenceString = content.substring(0, referenceIndex);</lineContent>` suppression stays, it means that anontation is not working.
I would rather remove annotation  a keep conversation in check issue.

romani replied:
@zyadahmed , please reopen issue to share details that it is not working even it supposed to work

zyadahmed replied:
> @zyadahmed , please reopen issue to share details that it is not working even it supposed to work

I will try to investigate more into suppressions and open issues and PRs of  currently exists issues is that what you mean  right ?

romani replied:
Yes, please.

romani reviewed (changes requested):
thanks a lot:

few more requests:

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -423,6 +441,100 @@ private static String topLevelType(String type) {
+     * @param tag the input tag to check
+     * @return a list of extracted references
+     */
+    public static List<JavadocTag> bestTryToMatchReference(JavadocTag tag) {
+        final String content = tag.getFirstArg();
+        final List<JavadocTag> validTags = new ArrayList<>();
```
why single tag mapped to list of tags.
match should return Optional, outside of this method , there should be a code that convert optional to required type.

but if you do like `tags.addAll(bestTryToMatchReference(getValidReferenceTag(validTags));` you will get as parameter list of tags, and map them to result filtered list. so it will be better from design perspectve.

zyadahmed replied:
> why single tag mapped to list of tags. match should return Optional, outside of this method , there should be a code that convert optional to required type.
> 
> but if you do like `tags.addAll(bestTryToMatchReference(getValidReferenceTag(validTags));` you will get as parameter list of tags, and map them to result filtered list. so it will be better from design perspectve.

yes that correct sorry for that i choose to use Optional

romani reviewed (changes requested):
items:

zyadahmed commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtilTest.java:
```diff
@@ -179,6 +179,7 @@ public void testInvalidTags() {
             "/** @fake block",
             " * {@bogus inline}",
             " * {@link List valid}",
+
```
> please revert

done

zyadahmed commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -344,14 +356,20 @@ private void collectReferencesFromJavadoc(DetailAST ast) {
+        // Process BLOCK tags
+        tags.addAll(getValidReferenceTag(
+                getValidTags(textBlock, JavadocUtil.JavadocTagType.BLOCK),
+                JavadocUtil.JavadocTagType.BLOCK
```
> too nested. make it like:
> 
> ```
> validTags = getValidTags(textBlock, JavadocUtil.JavadocTagType.BLOCK),
>                 JavadocUtil.JavadocTagType.BLOCK)
>         tags.addAll(bestTryToMatchReference(getValidReferenceTag(validTags));
> ```

i remove nested but change a little bit to make ci run inspections green

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -423,6 +441,100 @@ private static String topLevelType(String type) {
+     * @param javadocTag type of tag to check
+     * @return a {@code List} of valid Javadoc tags based on the specified type
+     */
+    public static List<JavadocTag> getValidReferenceTag(
```
getValidReferenceTags

as it return collection.

zyadahmed replied:
done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -423,6 +441,100 @@ private static String topLevelType(String type) {
+            Iterable<JavadocTag> tags, JavadocUtil.JavadocTagType javadocTag) {
+        final List<JavadocTag> validTags = new ArrayList<>();
+
+        for (JavadocTag tag : tags) {
+            if (javadocTag == JavadocUtil.JavadocTagType.BLOCK && tag.isInlineTag()) {
+                continue;
+            }
+            bestTryToMatchReference(tag).ifPresent(validTags::add);
+        }
```
try to rewrite is by java streams.
.filter().map().collect()

zyadahmed replied:
done with streams

romani reviewed (changes requested):
please generate diff report 
see details at https://github.com/checkstyle/checkstyle/tree/master/.github/workflows#diff-report-by-configuration-in-input-file

you can find numerous examples of usage in our old merged PRs

items:

zyadahmed commented:
Github, generate report for UnusedImport/all-examples-in-one

zyadahmed commented:
Github, generate report for UnusedImports/all-examples-in-one

zyadahmed commented:
> Report for UnusedImports/all-examples-in-one: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c5f69a32321cd7613369167c1be1c7193ab411c6_2025173059/reports/diff/index.html

The only difference in Checkstyle is that `InputUnusedImportsFromStaticMethodRefJavadocDisabled` generates a violation. It should already be a violation as I understand it. This is my first time using that report.

romani commented:
@mahfouz72 , please finish review, PR is on final stages

romani reviewed (changes requested):
Items

mahfouz72 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -343,12 +356,17 @@ private void collectReferencesFromJavadoc(DetailAST ast) {
-        // gather all the block-level tags, like @throws and @see
-        tags.addAll(getValidTags(textBlock, JavadocUtil.JavadocTagType.BLOCK));
+        // Process INLINE tags
+        final List<JavadocTag> inlineTags = getValidTags(textBlock,
+                                        JavadocUtil.JavadocTagType.INLINE);
+        final List<JavadocTag> tags = new ArrayList<>(getValidReferenceTags(inlineTags,
+                                                    JavadocUtil.JavadocTagType.INLINE));
+
+        // Process BLOCK tags
+        final List<JavadocTag> blockTags = getValidTags(textBlock,
+                                        JavadocUtil.JavadocTagType.BLOCK);
+        tags.addAll(getValidReferenceTags(blockTags,
+                                        JavadocUtil.JavadocTagType.BLOCK));
```
```suggestion
        final List<JavadocTag> validReferenceTags = new ArrayList<>();
        
        final List<JavadocTag> inlineTags = getValidTags(textBlock,
                                        JavadocUtil.JavadocTagType.INLINE);
        validReferenceTags.addAll(getValidReferenceTags(inlineTags, JavadocUtil.JavadocTagType.INLINE));

        final List<JavadocTag> blockTags = getValidTags(textBlock,
                                        JavadocUtil.JavadocTagType.BLOCK);
        validReferenceTags.addAll(getValidReferenceTags(blockTags,
                                        JavadocUtil.JavadocTagType.BLOCK));
```

zyadahmed replied:
> 

 tried to create that solution, but it failed during the inspection because it requires a parameterized constructor . another solution If I only declare final List<JavadocTag> validReferenceTags; and then add it in inline tags and block tags, it also fails the inspection, but with a 'Too far declaration' error."

mahfouz72 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -423,6 +441,99 @@ private static String topLevelType(String type) {
+     * @return a {@code List} of valid Javadoc tags based on the specified type
+     */
+    public static List<JavadocTag> getValidReferenceTags(
+            Collection<JavadocTag> tags, JavadocUtil.JavadocTagType javadocTag) {
```
```suggestion
            Collection<JavadocTag> tags, JavadocUtil.JavadocTagType javadocTagType) {
```

zyadahmed replied:
> 

done

mahfouz72 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -423,6 +441,99 @@ private static String topLevelType(String type) {
+        final List<JavadocTag> validTags = new ArrayList<>();
+
+        tags.stream()
+            .filter(tag -> !(javadocTag == JavadocUtil.JavadocTagType.BLOCK && tag.isInlineTag()))
```
Does this line mean that tags intended to be inline but written as block tags are ignored?

Example:
```
/**
  *   @link List
  */
```

For the above Javadoc:

- `javadocTagType == JavadocUtil.JavadocTagType.BLOCK` returns true because `@link` is written as a block tag.
- `tag.isInlineTag()` returns true because `@link` is an inline tag according to the Javadoc specification.

mahfouz72 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -423,6 +441,99 @@ private static String topLevelType(String type) {
+            }
+        }
+
+        final int methodIndex;
```
We can initialize this with -1 and remove the else branch.

zyadahmed replied:
> We can initialize this with -1 and remove the else branch.

done

mahfouz72 reviewed (changes requested):
Good Stuff!

Few items:

mahfouz72 commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/unusedimports/InputUnusedImportsWithLinkAndMethodTag.java:
```diff
@@ -0,0 +1,51 @@
+
+
+/**
+ * @link {Collections::emptyEnumeration}
+ * {@link TreeSet}
+ * {@link List#size}
```
All test cases for link and see tags only. Please add test cases for all tags that can reference imports :

```
    public boolean canReferenceImports() {
        return tagInfo == JavadocTagInfo.SEE
                || tagInfo == JavadocTagInfo.LINK
                || tagInfo == JavadocTagInfo.VALUE
                || tagInfo == JavadocTagInfo.LINKPLAIN
                || tagInfo == JavadocTagInfo.THROWS
                || tagInfo == JavadocTagInfo.EXCEPTION;
    }
```

Also, cases when the tag has a description. like : ` {@link List List of items}`

zyadahmed replied:
> All test cases for link and see tags only. Please add test cases for all tags that can reference imports :
> 
> ```
>     public boolean canReferenceImports() {
>         return tagInfo == JavadocTagInfo.SEE
>                 || tagInfo == JavadocTagInfo.LINK
>                 || tagInfo == JavadocTagInfo.VALUE
>                 || tagInfo == JavadocTagInfo.LINKPLAIN
>                 || tagInfo == JavadocTagInfo.THROWS
>                 || tagInfo == JavadocTagInfo.EXCEPTION;
>     }
> ```
> 
> Also, cases when the tag has a description. like : ` {@link List List of items}`

done in `InputUnusedImportsWithLinkTag`

zyadahmed commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -343,12 +356,17 @@ private void collectReferencesFromJavadoc(DetailAST ast) {
+        // Process INLINE tags
+        final List<JavadocTag> inlineTags = getValidTags(textBlock,
+                                        JavadocUtil.JavadocTagType.INLINE);
+        final List<JavadocTag> tags = new ArrayList<>(getValidReferenceTags(inlineTags,
```
> To let code looks same as next block

to make them both identical cli will fail in run inspection in 
```
<package>com.puppycrawl.tools.checkstyle.checks.imports</package>
<entry_point TYPE="method" FQNAME="com.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck java.util.Set<java.lang.String> collectReferencesFromJavadoc(com.puppycrawl.tools.checkstyle.api.TextBlock textBlock)"/>
<problem_class id="CollectionAddAllCanBeReplacedWithConstructor" severity="ERROR" attribute_key="ERRORS_ATTRIBUTES">Redundant 'Collection.addAll()' call</problem_class>
<description>'addAll()' call can be replaced with parametrized constructor call</description>
<highlighted_element>addAll</highlighted_element>
```

zyadahmed replied:
> ```
> final List<JavadocTag> tags = new ArrayList<>();
> // Process INLINE tags
> final List<JavadocTag> inlineTags = 
>         getValidTags(textBlock, JavadocUtil.JavadocTagType.INLINE);
> tags.addAll(getValidReferenceTags(inlineTags, JavadocUtil.JavadocTagType.INLINE));
> ```

i tried this and some similar it always fail in run inspections

romani replied:
sad, lets avoid violation.

please format code like:
```

        // Process INLINE tags
        final List<JavadocTag> inlineTags = 
                getValidTags(textBlock, JavadocUtil.JavadocTagType.INLINE);
        final List<JavadocTag> validTags = new ArrayList<>(
                getValidReferenceTags(inlineTags, JavadocUtil.JavadocTagType.INLINE));

        // Process BLOCK tags
        final List<JavadocTag> blockTags = 
                getValidTags(textBlock, JavadocUtil.JavadocTagType.BLOCK);
        validTags.addAll(
                getValidReferenceTags(blockTags, JavadocUtil.JavadocTagType.BLOCK));
```

romani commented:
@zyadahmed, please reply "done" to items you addressed.

Please keep CI green.

zyadahmed commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -423,6 +441,99 @@ private static String topLevelType(String type) {
+            .map(UnusedImportsCheck::bestTryToMatchReference)
+            .flatMap(Optional::stream)
+            .forEach(validTags::add);
+        return validTags;
```
> Try to optimize it more to get form:
> 
> ```
> return tags.stream().....
> ```

done

mahfouz72 commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/unusedimports/InputUnusedImportsWithLinkAndMethodTagSuperClass.java:
Why did we add this file?

zyadahmed replied:
i made this file to simulate the exceution of checkstyle dependecies where we have superclass that have some method and reference is mentioned inside subclass  which refer to superclass method and it's references  so i needed to make test same exactly as dependecies

mahfouz72 commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/unusedimports/InputUnusedImportsWithLinkTag.java:
```diff
@@ -0,0 +1,20 @@
+
+/**
+ * @link List
+ * {@link Set Set of items}
```
Another example like this but with invalid syntax:
`@link Set Set of items`  
and show that it works by having an unused import violation on `Set`

zyadahmed replied:
> Another example like this but with invalid syntax: `@link Set Set of items` and show that it works by having an unused import violation on `Set`

done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -423,6 +443,94 @@ private static String topLevelType(String type) {
+            Collection<JavadocTag> tags, JavadocUtil.JavadocTagType javadocTagType) {
+
+        return tags.stream()
+            .filter(tag -> javadocTagType != JavadocUtil.JavadocTagType.BLOCK || !tag.isInlineTag())
```
this condition is supper confusing.

we hardcoding value of javadocTagType 
and filtering by it, some nonsense is there.
```
       final List<JavadocTag> blockTags = getValidTags(textBlock,
                JavadocUtil.JavadocTagType.BLOCK);
        validTags.addAll(getValidReferenceTags(blockTags,
                JavadocUtil.JavadocTagType.BLOCK));
```

you should move it filtering to this method:
```
private static List<JavadocTag> getValidTags(TextBlock cmt,
            JavadocUtil.JavadocTagType javadocTagType) {
        return JavadocUtil.getJavadocTags(cmt, javadocTagType).getValidTags();
    }
```

so list of tags will be valid and we do not need not trust it.

romani reviewed (changes requested):
items:

mahfouz72 reviewed (changes requested):
Thanks alot!!

Last for me:

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -343,16 +357,22 @@ private void collectReferencesFromJavadoc(DetailAST ast) {
+        final List<JavadocTag> blockTags =
+                getValidTags(textBlock, JavadocUtil.JavadocTagType.BLOCK);
+        validTags.addAll(
+                getValidReferenceTags(blockTags));
```
here we combine two collections.
please use this form of code:
```
List<JavadocTag> targetTags = Stream
                .concat(listxxxx1.stream(), 
                       listxxxxxx2.stream())
                .collect(Collectors.toList());
```

romani reviewed (changes requested):
we are very close:

zyadahmed commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -362,12 +382,16 @@ private static Set<String> collectReferencesFromJavadoc(TextBlock textBlock) {
+        return JavadocUtil.getJavadocTags(cmt, javadocTagType)
+            .getValidTags()
+            .stream()
+            .filter(tag -> javadocTagType != JavadocUtil.JavadocTagType.BLOCK || !tag.isInlineTag())
```
> `.filter(tag -> javadocTagType != JavadocUtil.JavadocTagType.BLOCK || !tag.isInlineTag())`
> 
> this condition is weird. `javadocTagType != JavadocUtil.JavadocTagType.BLOCK` doe snot use tag.
> 
> sounds like we just need separate method for BLOCK

This filter is more related to whether the tag type matches or differs from the extraction format. For example, `@link `is an inline tag but is extracted as a block tag. I tried different ways to simplify it, but maybe we can move it to a separate method and explain it properly in the Javadoc summary. What do you think?

romani replied:
Please try.

zyadahmed replied:
> add ` .map(UnusedImportsCheck::bestTryToMatchReference)` to this chain.
> 
> and we do not need `getValidReferenceTags(`

done

zyadahmed replied:
> Please try.

i tried to make simplest conditon as much as possible and write detailed javadoc about usage

zyadahmed commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -343,16 +357,22 @@ private void collectReferencesFromJavadoc(DetailAST ast) {
+        // Process INLINE tags
+        final List<JavadocTag> inlineTags =
+                getValidTags(textBlock, JavadocUtil.JavadocTagType.INLINE);
+        final List<JavadocTag> validTags = new ArrayList<>(
```
> targetTags

done

zyadahmed commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java:
```diff
@@ -362,12 +382,16 @@ private static Set<String> collectReferencesFromJavadoc(TextBlock textBlock) {
+     * @param javadocTagType The type of tags we're interested in
      * @return the list of tags
      */
     private static List<JavadocTag> getValidTags(TextBlock cmt,
```
> `getTargetTags`
> 
> as we filter out completely valid tags only because they are not our interest. Please update javadoc summary to describe that what we return or what we filterout.

done

romani commented:
@mahfouz72 , please do final review , please share your opinion on possible extra annoation `@Regexp` in code to avoid checker to not place violation.

zyadahmed commented:
Github, generate report for UnusedImports/all-examples-in-one

zyadahmed commented:
Github, generate report for UnusedImports/all-examples-in-one

romani commented:
Rebased in web

romani reviewed (approved):
Thanks a lot for update,
I am approving to let one more maintainer to  review while you finishing with checkerwramework, please try to avoid Checker suppression.

romani commented:
@zyadahmed, if annotation are not working, please revert usage of them, and share feedback in Checker issue tracker, we will merge with suppression

zyadahmed commented:
> @zyadahmed, if annotation are not working, please revert usage of them, and share feedback in Checker issue tracker, we will merge with suppression

revert last update and opened pr  https://github.com/typetools/checker-framework/issues/7013

zyadahmed commented:
> @zyadahmed, if annotation are not working, please revert usage of them, and share feedback in Checker issue tracker, we will merge with suppression

revert last update and opened pr  https://github.com/typetools/checker-framework/issues/7013

zyadahmed commented:
Github, generate report for UnusedImports/all-examples-in-one

romani commented:
Github, generate report for UnusedImports/all-examples-in-one

romani commented on config/import-control.xml:
```diff
@@ -184,6 +184,7 @@
     </file>
     <subpackage name="imports">
       <allow class="com.puppycrawl.tools.checkstyle.XmlLoader" local-only="true"/>
+      <allow class="org.checkerframework.checker.index.qual.IndexOrLow"/>
```
Please remove, they are not working.

zyadahmed replied:
> Please remove, they are not working.

But after we added this annotation, we removed one suppression, or am I missing something?

```
 <message>incompatible argument for parameter endIndex of String.substring.</message>
    <lineContent>referenceString = content.substring(0, referenceIndex);</lineContent>
    <details>
      found   : int
      required: @NonNegative int
   </details>
```

romani replied:
I saw red CI on Checker .... that is why I requested to remove.

I am very confused. Ok lets merge with suppressions and improve in small PR on suppression removal

romani replied:
Oh my bad, it actually helped but another poped up:
```

  <checkerFrameworkError unstable="false">
    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java</fileName>
    <specifier>return</specifier>
    <message>incompatible types in return.</message>
    <lineContent>return methodIndex;</lineContent>
    <details>
      type of expression: int
      method return type: @LTLengthOf(&quot;input&quot;) int
    </details>
  </checkerFrameworkError>
```

ok, lets keep this annotation. And resolve above new in separate PR.
